components/zoho_crm/sources/common/http-based/base.js

```diff
@@ -0,0 +1,280 @@
+const {
+  randomBytes,
+  randomInt,
+} = require("crypto");
+const zoho_crm = require("../../../zoho_crm.app");
+
+// Zoho CRM webhooks subscriptions have an expiration date of up to 1 day. This
+// event source renews the subscription every 12 hours by default. More info can
+// be found in the developer docs:
+// https://www.zoho.com/crm/developer/docs/api/v2/notifications/enable.html
+const hookRenewalPeriod = 60 * 60 * 12;
+
+module.exports = {
+  dedupe: "unique",
+  props: {
+    zoho_crm,
+    db: "$.service.db",
+    http: "$.interface.http",
+    timer: {
+      type: "$.interface.timer",
+      default: {
+        intervalSeconds: hookRenewalPeriod,
+      },
+    },
+  },
+  hooks: {
+    async activate() {
+      const token = this._generateToken();
+      const channelId = this._getRandomChannelId();
+      const channelExpiry = this._getChannelNextExpiryDate();
+      const events = this.getEvents();
+      const hookOpts = {
+        token,
+        channelId,
+        notifyUrl: this.http.endpoint,
+        channelExpiry,
+        events,
+      };
+
+      await this.zoho_crm.createHook(hookOpts);
+
+      console.log(`
+        Created webhook notification for channel ID: ${channelId}
+      `);
+
+      this.db.set("token", token);
+      this.db.set("channelId", channelId);
+    },
+    async deactivate() {
+      const channelId = this.db.get("channelId");
+      await this.zoho_crm.deleteHook(channelId);
+
+      console.log(`
+        Deleted webhook notification for channel ID: ${channelId}
+      `);
+
+      this.db.set("token", null);
+      this.db.set("channelId", null);
+    },
+  },
+  methods: {
+    _generateToken() {
+      // The max size of a verification token is 50 chars:
+      // https://www.zoho.com/crm/developer/docs/api/v2/notifications/enable.html
+      const tokenSize = 49;
+      return randomBytes(tokenSize)
+        .toString("base64")
+        .slice(0, tokenSize);
+    },
+    _getChannelNextExpiryDate() {
+      const currentTimestamp = Date.now();
+      const nextTimestamp = currentTimestamp + hookRenewalPeriod * 1000;
+      const nextExpiryDate = new Date(nextTimestamp);
+      return (
+        nextExpiryDate
+          .toISOString()
+          // The API does not accept date ISO strings that include milliseconds
+          .replace(/\.\d{3}Z$/, "Z")
```

Comment:
Will an ISO 8601 ms portion always have exactly 3 digits? Just wanted to make sure this would match all conditions.

Reply:
Yes, it always ends in `.sssZ`
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toISOString

Regardless, this call should be safe enough:
```js
> const dateStr = new Date('2021-05-28 15:15:16.1234 PST').toISOString()
undefined
> dateStr.slice(0, -5)
'2021-05-28T23:15:16'
> dateStr.slice(0, -5).replace(/\.\d{3}Z$/, "Z")
'2021-05-28T23:15:16'
```